codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -393,21 +413,27 @@ public ChannelFuture writeWindowUpdate(ChannelHandlerContext ctx, int streamId,
     @Override
     public ChannelFuture writeFrame(ChannelHandlerContext ctx, byte frameType, int streamId,
             Http2Flags flags, ByteBuf payload, ChannelPromise promise) {
-        boolean releaseData = true;
         SimpleChannelPromiseAggregator promiseAggregator =
                 new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
         try {
             verifyStreamOrConnectionId(streamId, STREAM_ID);
             ByteBuf buf = ctx.alloc().buffer(FRAME_HEADER_LENGTH);
+            // Assume nothing below will throw until buf is written. That way we don't have to take care of ownership
+            // in the catch block.
             writeFrameHeaderInternal(buf, payload.readableBytes(), frameType, flags, streamId);
             ctx.write(buf, promiseAggregator.newPromise());
-
-            releaseData = false;
-            ctx.write(payload, promiseAggregator.newPromise());
         } catch (Throwable t) {
-            if (releaseData) {
+            try {
                 payload.release();
+            } finally {
+                promiseAggregator.setFailure(t);
+                promiseAggregator.doneAllocatingPromises();
             }
+            return promiseAggregator;
+        }
+        try {
+            ctx.write(payload, promiseAggregator.newPromise());
```

Comment:
should this be a voidPromise?

Reply:
can you explain why you think this? We don't want `promise` to be completed until the 2 writes in this method (frame header, and payload) are complete. `SimpleChannelPromiseAggregator` facilitates the combination of multiple promises events into a single promise notification, but it needs to know how many promise events to expect. `newPromise` actually just increments a counter and `doneAllocatingPromises` signifies the counter value is completed and notification is now allowed.